### ISSUE TITLE
search timed out, try again or try another indexer

### ISSUE THREAD

**Issue by dpeart** [SiCKRAGETV/sickrage-issues#563]
When I try to add new shows, I get the error;
     search timed out, try again or try another indexer

I turned on debug and only get this in the log file:

     2015-02-07 13:56:23 DEBUG    Thread-13 :: Searching for Show with searchterm: The This Old House Hour on Indexer: theTVDB

Python version: 2.7.5 (default, May 29 2013, 02:28:51) [GCC 4.8.0]
Slackware

thanks,
david

**Comment by dpeart** [comment]
I do have network access to theTVDB.com as I can wget it and ping it.

**Comment by abeloin** [comment]
The tvdb server is having some issue today, seen multiples cloudfare timeout today.
There's not much we can do, you have to be patient and retry.

**Comment by dpeart** [comment]
I've been having this issue for about 4 days, neither tvdb nor tvrage have been working.

I'll keep trying and see if it changes anytime soon.

**Comment by abeloin** [comment]
Then maybe something else is going on, your sickrage branch/commit?

On the pc where sickrage is installed, wget work?

**Comment by dpeart** [comment]
wget works I've tried both main and develop

I wish I could get more debug message, there just aren't any

**Comment by abeloin** [comment]
Do you use a proxy for the indexers? Check in General Configuration -> Advanced Settings

**Comment by dpeart** [comment]
The proxy host portion was empty, but the check box was checked, so I uncheck and it made no difference.  I then increase my time out from 20 to 30 and still no luck.

I checked wget tvdb.org and it does get the index.html page so that does appear to work.

**Comment by abeloin** [comment]
Do you have debug=1 in config.ini? You will need to stop sickrage before changing the setting. Then you should have more info(e.g.: Connection timed out ... while loading URL)

You could try this test.py, if you think it's python related:
```
import requests
from requests.packages.urllib3 import disable_warnings
url = 'http://thetvdb.com/api/0629B785CE550C8D/series/73388/en.xml'
session = requests.Session()
response = session.get(url, timeout=30, verify=False, allow_redirects=True)
print response.headers
print response.content
```

**Comment by abeloin** [comment]
By the way what's your branch/commit?

**Comment by dpeart** [comment]
BRANCH: (develop) / COMMIT: (c62d8f2c799edd7daad5648f45276a3464d83a9e)

**Comment by abeloin** [comment]
You can add in sickbeard/logger.py at line 73 in self.loggers:
```logging.getLogger("requests.packages.urllib3")```

you should see in the the result of every request(e.g.: tvdb search):
```2015-02-07 18:36:01 DEBUG    "GET /api/GetSeries.php?seriesname=detective+conan&language=en HTTP/1.1" 200 None```

**Comment by dpeart** [comment]
I tried the test.py, but just get a ImportError: No module named requests.

**Comment by dpeart** [comment]
I did this, but it didn't work:

class SickRage(object):
    def __init__(self):
        logging.getLogger("requests.packages.urllib3")
        # system event callback for shutdown/restart
        sickbeard.events = Events(self.shutdown)

Sorry I'm not familiar with python.

david

**Comment by abeloin** [comment]
Lib requests is missing:
If you have pip:
```pip install requests```
If not you'll have to install it: 
* Download: https://bootstrap.pypa.io/get-pip.py
* Command line: # python get-pip.py

**Comment by abeloin** [comment]
It won't work your editing the wrong file, it's sickbeard/logger.py not SickBeard.py ;)

**Comment by dpeart** [comment]
Figure that out. :)

Here is all I get:
20:26:57 DEBUG::Thread-13 :: Searching for Show with searchterm: how it's made on Indexer: theTVDB
20:26:57 INFO::Starting new HTTP connection (1): thetvdb.com
20:26:57 INFO::Starting new HTTP connection (1): thetvdb.com
20:26:57 DEBUG::"GET /api/GetSeries.php?seriesname=how+it%27s+made&language=en HTTP/1.1" 200 None
20:26:57 DEBUG::"GET /api/GetSeries.php?seriesname=how+it%27s+made&language=en HTTP/1.1" 200 None


Then I tried a wget using that string and it worked.  I got back the proper thing.

hmmm

**Comment by abeloin** [comment]
Weird, you get a 200(ok) which mean it's successful.

**Comment by dpeart** [comment]
When I run TVRage I get:
20:32:47 INFO::Starting new HTTP connection (1): services.tvrage.com
20:32:47 INFO::Starting new HTTP connection (1): services.tvrage.com
20:33:04 DEBUG::"GET /feeds/search.php?show=how+it%27s+made HTTP/1.1" 200 1386
20:33:04 DEBUG::"GET /feeds/search.php?show=how+it%27s+made HTTP/1.1" 200 1386

**Comment by abeloin** [comment]
Modified anything in sickrage's code?
Also do you see some ```Connection timed out``` in the log

**Comment by dpeart** [comment]
No I've made no modifications, a clean check out.

I don't see any such messages.

**Comment by abeloin** [comment]
Can you send me a copy of the log on pastebin?

**Comment by dpeart** [comment]
sure give me a sec

**Comment by dpeart** [comment]
ok try this:

http://pastebin.com/CbkNCZVi

**Comment by abeloin** [comment]
Have you tried searching 'how it's made' with tvdb only?

**Comment by dpeart** [comment]
yes, I get the same output as I do with 'all indexers' is selected:
2015-02-07 21:44:19 DEBUG    Thread-62 :: Searching for Show with searchterm: how it's made on Indexer: theTVDB
2015-02-07 21:44:19 INFO     Starting new HTTP connection (1): thetvdb.com
2015-02-07 21:44:19 INFO     Starting new HTTP connection (1): thetvdb.com
2015-02-07 21:44:20 DEBUG    "GET /api/GetSeries.php?seriesname=how+it%27s+made&language=en HTTP/1.1" 200 None
2015-02-07 21:44:20 DEBUG    "GET /api/GetSeries.php?seriesname=how+it%27s+made&language=en HTTP/1.1" 200 None

**Comment by abeloin** [comment]
If you do a ```git dif``` what's the result?

**Comment by dpeart** [comment]
Only the logger change I made

diff --git a/sickbeard/logger.py b/sickbeard/logger.py
index 1d289f1..a82c4e3 100644
--- a/sickbeard/logger.py
+++ b/sickbeard/logger.py
@@ -70,6 +70,7 @@ class Logger(object):
             logging.getLogger('sickrage'),
             logging.getLogger('tornado.general'),
             logging.getLogger('tornado.application'),
+            logging.getLogger("requests.packages.urllib3")
             # logging.getLogger('tornado.access'),
         ]

**Comment by abeloin** [comment]
Mmm... a weird one.
Have you tried it on a fresh copy of sickrage(new config)?

**Comment by dpeart** [comment]
This is a new one from this AM.

**Comment by dpeart** [comment]
Ok I just moved that install aside and did a new git, using master branch this time.  I get the same results:

BRANCH: (master) / COMMIT: (401cb666016e45b42fe675bafdf7908ad6c2b9bb)

**Comment by abeloin** [comment]
Ok, can you try this modified script(use sickrage requests)? Place it in the root of sickrage path. 
test.py
```
import os, sys
sys.path.insert(1, os.path.abspath(os.path.join(os.path.dirname(__file__), 'lib')))
import requests
url = 'http://thetvdb.com/api/0629B785CE550C8D/series/73388/en.xml'
session = requests.Session()
response = session.get(url, timeout=30, verify=False, allow_redirects=True)
print response.history
print response.headers
print response.content

**Comment by dpeart** [comment]
I am using Python 2.7.5 for x64.  When I had it working before I was runnin Python 2.7.3 x32.  When I updated to x64 I used the new version.

**Comment by abeloin** [comment]
Should not be an issue as I'm running on 2.7.5 on x64 machine:
```
Python 2.7.5 (default, Nov  3 2014, 14:26:24) 
[GCC 4.8.3 20140911 (Red Hat 4.8.3-7)] on linux2
```

**Comment by dpeart** [comment]
yep, I installed 2.7.3 and you're right it didn't make any difference

**Comment by dpeart** [comment]
So I just installed sickbeard and it works without issues.  Just cloned it in.  Too bad it doesn't do torrents like sickrage :)

**Comment by abeloin** [comment]
Have you tried the script?

It is weird that you get this when the connection is succesful:
[Imgur](http://i.imgur.com/MZ3cNhb.png)

**Comment by dpeart** [comment]
I don't have pip

**Comment by abeloin** [comment]
The second script doesn't need pip, it use sickrage's requests lib.

**Comment by dpeart** [comment]
root@Tower:/mnt/user/apps/sickbeard_v1# python test.py
[]
CaseInsensitiveDict({'content-encoding': 'gzip', 'transfer-encoding': 'chunked', 'set-cookie': '__cfduid=d717e65f9b79610e0827f6e76a30aeaae1423371900; expires=Mon, 08-Feb-16 05:05:00 GMT; path=/; domain=.thetvdb.com; HttpOnly', 'cf-cache-status': 'REVALIDATED', 'expires': 'Sun, 08 Feb 2015 05:35:00 GMT', 'vary': 'Accept-Encoding', 'server': 'cloudflare-nginx', 'last-modified': 'Sat, 07 Feb 2015 07:13:18 GMT', 'connection': 'keep-alive', 'cache-control': 'public, max-age=1800', 'date': 'Sun, 08 Feb 2015 05:05:00 GMT', 'cf-ray': '1b554a2bfdd2095c-DFW', 'content-type': 'text/xml'})
<?xml version="1.0" encoding="UTF-8" ?>
<Data>
  <Series>
    <id>73388</id>
    <Actors>|Jamie Hyneman|Adam Savage|Kari Byron|Grant Imahara|Tory Belleci|Scottie Chapman|Heather Joseph-Witham|Robert Lee|Jessi Combs|Christine Chamberlain|</Actors>
    <Airs_DayOfWeek>Saturday</Airs_DayOfWeek>
    <Airs_Time>9:00 PM</Airs_Time>
    <ContentRating>TV-PG</ContentRating>
    <FirstAired>2003-01-23</FirstAired>
    <Genre>|Documentary|Reality|Special Interest|</Genre>
    <IMDB_ID>tt0383126</IMDB_ID>
    <Language>en</Language>
    <Network>Discovery</Network>
    <NetworkID></NetworkID>
    <Overview>So exactly how hard is it to find a needle in a haystack, anyway? And can water dripping on your forehead really drive you nuts? Those are the kinds of questions, myths and urban legends that are put to the test in this humorous series to find out which ones hold water (and keep it from dripping on your forehead and driving you insane).</Overview>
    <Rating>9.3</Rating>
    <RatingCount>268</RatingCount>
    <Runtime>60</Runtime>
    <SeriesID>22839</SeriesID>
    <SeriesName>MythBusters</SeriesName>
    <Status>Continuing</Status>
    <added></added>
    <addedBy></addedBy>
    <banner>graphical/73388-g10.jpg</banner>
    <fanart>fanart/original/73388-15.jpg</fanart>
    <lastupdated>1423292643</lastupdated>
    <poster>posters/73388-2.jpg</poster>
    <tms_wanted_old>1</tms_wanted_old>
    <zap2it_id>SH00557678</zap2it_id>
  </Series>
</Data>

**Comment by abeloin** [comment]
ok, can you try this in your web browser(change the host to point to your sickrage host):

http://localhost:8081/home/addShows/searchIndexersForShowName?search_term=how%20it%27s%20made&lang=en&indexer=1

**Comment by dpeart** [comment]
I see the debug message in the log file, but it never returns.  The web browser spins forever.

**Comment by abeloin** [comment]
Can you try it with wget?

**Comment by dpeart** [comment]
Hangs, but get same message i the log file, that it is search for the show.

 wget "http://localhost:8081/home/addShows/searchIndexersForShowName?search_term=how%20it%27s%20made&lang=en&indexer=1"
--2015-02-07 23:27:35--  http://localhost:8081/home/addShows/searchIndexersForShowName?search_term=how%20it%27s%20made&lang=en&indexer=1
Resolving localhost (localhost)... 127.0.0.1
Connecting to localhost (localhost)|127.0.0.1|:8081... connected.
HTTP request sent, awaiting response...

**Comment by abeloin** [comment]
If you want you can try debugging using python's pdb:

* Insert ```import pdb; pdb.set_trace()``` in line 2076 of sickbeard/webserve.py:
```
            t = sickbeard.indexerApi(indexer).indexer(**lINDEXER_API_PARMS)
            import pdb; pdb.set_trace()
            logger.log("Searching for Show with searchterm: %s on Indexer: %s" % (
```
* Start sickrage manually from the command line with the params specified in your startup script(if you datadir or like add to command line):
```
python /path/to/SickBeard.py --nolaunch -q
```
* Add a show in Sickrage.
* Go back to the command line, you should see something like this:
```
> .../webserve.py(2077)searchIndexersForShowName()
->
```
* If it's the case, press n then enter.
* Wait until the ```->``` is shown again.
* Repeat until it freeze.
* Take note of the line where it stuck
* Quit sickrage: Ctrl+C
* Restart sickrage like in the previous step
* At the line were it stuck, type s instead of n.
* Repeat until it freeze.
* Copy paste from the command line window.
* Post the result on pastebin

More info python debugging: https://pythonconquerstheuniverse.wordpress.com/2009/09/10/debugging-in-python/

**Comment by dpeart** [comment]
OK I'll do this tomorrow.  Kind of late for me tonight.

Thanks for all the help.

**Comment by abeloin** [comment]
Good night

**Comment by dpeart** [comment]
Ok I can't get through past here:
                results.setdefault(indexer, []).extend(t[search_term])
That is where it hangs, about three lines past where I turned on the debugger.  I've stepped inside it, but after about 100  steps it still didn't hang.  I'll continue stepping through.

**Comment by abeloin** [comment]
Your maybe in a loop, press l (lowercase L) and post the data here

**Comment by abeloin** [comment]
Forgot to mention, you can use r also which will run the program until it it a return statement.

**Comment by dpeart** [comment]
r would be nice, is there a way to do 32 nexts?   Here is where I am:

3xn
 results.setdefault(indexer, []).extend(t[search_term])
s
4xn
selected_series = self._getSeries(key)
s
n
allSeries = self.search(series)
s
4xn
 return self._getetsrc(self.config['url_getSeries'], self.config['params_getSeries']).values()[0]
s
2xn
return self._loadUrl(url, params=params, language=language).values()[0]
s
4xn
 return f(*args, **kwargs)
s
6xn
resp = session.get(url.strip(), cache_auto=True, params=params)
s
8xn
 return super(CacheControlSession, self).get(*args, **kw)
s
2xn
return self.request('GET', url, **kwargs)
s
32xn
 resp = self.send(prep, **send_kwargs)
s
15xn
r = adapter.send(request, **kwargs)
s
5xn
resp = super(CacheControlAdapter, self).send(request, **kw)

**Comment by abeloin** [comment]
Use markdown to wrap the result like code, it will be more easier to read. If you press enter without any character it reuse the previous one.

**Comment by abeloin** [comment]
By the way, at ```results.setdefault(indexer, []).extend(t[search_term])``` press s then at the next(selected_series = self._getSeries(key)) press r and repeat it should go faster.

**Comment by dpeart** [comment]
OK I found the loop.

If I run with:

b /mnt/user/apps/sickbeard_v1/lib/lockfile/linklockfile.py 29

That is in the middle of the endless loop.  Spinning on a file lock?

**Comment by dpeart** [comment]
nlinks returns a 1, it is looking for a 2 so it thinks the lock failed and re-runs.

**Comment by abeloin** [comment]
What's the value of self.unique_name?
```p self.unique_name```

**Comment by abeloin** [comment]
And also self.lock_file?

**Comment by abeloin** [comment]
I'm flabbergasted there's no timeout by default!!!

Ok remove ```import pdb; pdb.set_trace()``` from sickbeard/webserve.py and put into lib/lockfile/linklockfile.py line 26. It should be more easier the get the value.

Permission issue?

**Comment by dpeart** [comment]
I can see the self.unique_name file in cache/sessions.  It exsits. However, when I run os.link(self.unique_name, self.lock_file) I get the following error:
(Pdb) > /mnt/user/apps/sickbeard_v1/lib/lockfile/linklockfile.py(30)acquire()
-> os.link(self.unique_name, self.lock_file)
s
(Pdb) OSError: (38, 'Function not implemented')
> /mnt/user/apps/sickbeard_v1/lib/lockfile/linklockfile.py(30)acquire()
-> os.link(self.unique_name, self.lock_file)

**Comment by dpeart** [comment]
I also just figured out that ln test foo gives me the same error.  For some reason I cannot create hard links.  I'm running this on unRaid (my NAS) box, I'll have to go and figure out why this is happening.  Interestingly enough sickbeard doesn't have this issue.  I'll have to pull sickbeard down and see how they implemented this.

**Comment by dpeart** [comment]
wow, I just learned something, so I guess this debug effort was good for me.  In unRaid they are "user" shares, and on user shares you cannot do a hard link.  Sickrage gets hung up trying to hard link.  I guess sickbeard doesn't.

I"m not sure if there is a way to make sickrage more tollerant of not being able to do a hard link, but I moved my sickrage install and it now works.

ableoin I cannot thank you enough for helping me figure this out.  And now I've run the python debugger :)

**Comment by abeloin** [comment]
The joy of debugging ;)
For the difference between sickbeard/sickrage, my guess is sickbeard don't use lockfile.
I'm currently looking for way to check if it's an unraid system.

**Comment by abeloin** [comment]
@dpeart Can you try this commit: abeloin/SickRage@de6cd8d9f1cb62f34f6a597f9578ecd0d11e9f3a I think it should fix the issue.

**Comment by dpeart** [comment]
how do I test that commit?

**Comment by abeloin** [comment]
* Stop sickrage
* Download the diff: https://github.com/abeloin/SickRage/commit/de6cd8d9f1cb62f34f6a597f9578ecd0d11e9f3a.diff
* In sickrage folder: ```git apply /path/to/de6cd8d9f1cb62f34f6a597f9578ecd0d11e9f3a.diff```
* Start sickrage
* Use the test url the previous comment

**Comment by dpeart** [comment]
that worked!

**Comment by abeloin** [comment]
Yes and thank you for the debugging.

@fernandog The pull request SiCKRAGETV/SickRage#1287 fix the issue.

**Comment by dpeart** [comment]
This is what I got back:
{"results": [["theTVDB", 1, "http://thetvdb.com/?tab=series&id=", 79485, "How It's Made", "2001-01-06"]], "langid": 7}

I then tried it in the "gui" and it worked there as well.

**Comment by abeloin** [comment]
Confirmed, I got the same results.

**Comment by fernandog** [comment]
Nice! Sorry I'm away for a while. New boss at work. At home my boss is my kid =)

**Comment by fernandog** [comment]
@markheloking can you help me labeling the new issues?

**Comment by abeloin** [comment]
No worries, kids are and always will be the priority no matter what. :)

**Comment by fernandog** [comment]
@dpeart develop updated. please test and report back

**Comment by abeloin** [comment]
@dpeart Any news?

**Comment by dpeart** [comment]
I am running:

BRANCH: (master) / COMMIT: (93f0f0ad1fa278afb17f73d47c099b82a766dede)

 

and it times out, does that commit (latest master) include the fix?

 

When I look at it lib/lockfile/linklockfile.py I don’t see the error code 38 in there.

 

david

**Comment by fernandog** [comment]
@dpeart no. only develop. change branch in general>advanced>github

**Comment by abeloin** [comment]
It was merged in master 3 days ago: https://github.com/SiCKRAGETV/SickRage/commits/master/lib/lockfile/linklockfile.py

**Comment by abeloin** [comment]
@dpeart Can you place a ```import pdb; pdb.set_trace()``` in https://github.com/SiCKRAGETV/SickRage/blob/master/lib/lockfile/linklockfile.py#L27 to see if the error is still 38?

**Comment by dpeart** [comment]
My linklockfile.py is not what you show here:

 

        while True:

            # Try and create a hard link to it.

            try:

                os.link(self.unique_name, self.lock_file)

            except OSError:

                # Link creation failed.  Maybe we've double-locked?

                nlinks = os.stat(self.unique_name).st_nlink

 

There is no:

 

if e.errno == errno.ENOSYS:

 

As seen in the github file linked below.

 

david

**Comment by dpeart** [comment]
Ok for some reason my docker hadn’t updated completely.

 

After updating, restarting and verifying the proper linklockfile.py was in place.  It worked just fine.

 

thanks

david

**Comment by abeloin** [comment]
Thank you for the testing

**Comment by bughatti** [comment]
team, I know this is closed but I wanted to update my findings.  I have Synology DSM DSM 5.2-5592 Update 1 on a DS 414.  I installed Sickrage through the synocommunity today and started to work on getting it setup.  I followed a few different sites to get it configured and when I tried to start searching, I received the dreaded search timed out, try again or try another indexer.  I looked at multiple git pages and sickragetv pages for fixes, yet none of them worked.  I noticed alot of questions had been asked if people had a proxy setup.  I did not in sickrage but I do have my DSM network configured to use a NL proxy.  I had enabled the debug and did a tail -f on the log when i searched and never received any errors.  After a few hours I decided to set the proxy on my search providers since I had a proxy on my DSM and at least TVrage started to pull data.  Every once in a while TheTVDB will pull data but its a hit and miss and no log failures for anything to suggest what I should change.  Like I said it is finally working so this is just info for others that want to use a proxy on the synology and still use sickrage
